Add EventCorrelatorOptions for event rate limiting

mrogers950 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
With the client side rate limiting from #47367 we also need to be able to create event sinks with non-default cache, aggregation, and spam settings.

**Which issue this PR fixes**:

**Special notes for your reviewer**:
depends on #47367

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```NONE
```
@derekwaynecarr

dims commented:
/ok-to-test

enj commented:
@smarterclayton @liggitt @derekwaynecarr @mml can we consider this for 1.8 now that https://github.com/kubernetes/kubernetes/pull/47367 is in the queue?

smarterclayton commented:
Please rebase and retest for 1.9

mikedanese commented on staging/src/k8s.io/client-go/tools/record/events_cache.go:
```diff
@@ -331,43 +416,73 @@ type EventCorrelateResult struct {
+	return NewEventCorrelatorOptions(maxLruCacheEntries, defaultSpamBurst, defaultSpamQPS, defaultAggregateMaxEvents, defaultAggregateIntervalInSeconds)
+}
+
+func NewEventCorrelatorOptions(cacheSize, spamBurst int, spamQPS float32, aggregateMax, aggregateInterval int) *EventCorrelatorOptions {
```
This function seems not useful and will have to change every time we add a filed to EventCorrelatorOptions

Commit:
Add EventCorrelatorOptions for changing event sink settings

Allows creation of an event sink with non-default spam and aggregate
parameters. Add tests that exercise this with a 1 event per minute limit.

Signed-off-by: Matt Rogers <mrogers@redhat.com>

mrogers950 commented:
@smarterclayton rebased.

enj commented:
/test all

simo5 commented:
@smarterclayton can you review this and approve ?

enj commented:
/lgtm

mrogers950 commented:
/test all

mrogers950 commented:
@smarterclayton , @derekwaynecarr does this PR need anything else before approval?

nikhita commented:
/retest

nikhita commented:
/retest
